experimental/traffic-portal/src/app/core/users/tenants/tenant-details/tenant-details.component.ts

```diff
@@ -0,0 +1,175 @@
+	styleUrls: ["./tenant-details.component.scss"],
+	templateUrl: "./tenant-details.component.html"
+})
+export class TenantDetailsComponent implements OnInit {
+	public new = false;
+	public disabled = false;
+	public tenant!: Tenant;
+	public tenants = new Array<ResponseTenant>();
+	public displayTenant: TreeData;
+
+	constructor(private readonly route: ActivatedRoute, private readonly userService: UserService,
+		private readonly location: Location) {
+		this.displayTenant = {
+			children: [],
+			id: -1,
+			name: ""
+		};
+	}
+
+	/**
+	 * Catches when tree-select outputs an update event
+	 *
+	 * @param evt The TreeData selected
+	 */
+	public update(evt: TreeData): void {
+		const tenant = this.tenants.find(t => t.id === evt.id);
+		if (tenant === undefined) {
+			console.error(`Unknown tenant selected ${evt.id}`);
+			return;
+		}
+		this.tenant.parentId = tenant.id;
+	}
+
+	/**
+	 * Recursively fills out a nodes children.
+	 *
+	 * @param tenantByParentId All tenants grouped by parent id.
+	 * @param currentTenant The tenant to populate.
+	 */
+	public breakTenantNode(tenantByParentId: Map<number, Array<ResponseTenant>>, currentTenant: TreeData): void {
+		currentTenant.children = (tenantByParentId.get(currentTenant.id) ?? []).map(t => ({...t, children: []} as TreeData));
+
+		currentTenant.children.forEach(t => {
+			this.breakTenantNode(tenantByParentId, t);
+		});
+	}
+
+	/**
+	 * Converts the tenants list into the tree-data structure needed by the tree-select component.
+	 */
+	public constructTreeData(): void {
+		const tenantByParentId = new Map<number, Array<ResponseTenant>>();
+		this.tenants.forEach(t => {
+			if (t.parentId === null) {
+				return;
+			}
+			if (!tenantByParentId.has(t.parentId)) {
+				tenantByParentId.set(t.parentId, []);
+			}
+			// The above check ensures this always returns at least `[]`
+			// @ts-ignore
+			tenantByParentId.get(t.parentId).push({...t, children: []});
+		});
+		const rootTenant = this.tenants.find(t => t.parentId === null);
+		if (rootTenant === undefined) {
+			return;
+		}
+		const rootNode = {...rootTenant, children: []} as TreeData;
+		this.breakTenantNode(tenantByParentId, rootNode);
+
+		this.displayTenant = rootNode;
+	}
+
+	/**
+	 * Angular lifecycle hook.
+	 */
+	public async ngOnInit(): Promise<void> {
+		const ID = this.route.snapshot.paramMap.get("id");
+		if (ID === null) {
+			console.error("missing required route parameter 'id'");
+			return;
+		}
+
+		this.tenants = await this.userService.getTenants();
+		this.constructTreeData();
+
+		if (ID === "new") {
+			this.new = true;
+			this.tenant = {
+				active: true,
+				name: "",
+			} as RequestTenant;
+			return;
+		}
+		const numID = parseInt(ID, 10);
+		if (Number.isNaN(numID)) {
+			console.error("route parameter 'id' was non-number:", ID);
+			return;
+		}
+		const tenant = this.tenants.find(t => t.id === numID);
+		if (!tenant) {
+			throw new Error(`Unable to find tenant with id ${numID}`);
```

Comment:
Why does this throw when the other errors are just `console.error`'d? I feel like we should pick one.